File: frontend/web/monorepo/libs/mateu/src/mateu/shared/apiClients/dtos/componentmetadata/Trigger.ts
import { TriggerType } from "@mateu/shared/apiClients/dtos/componentmetadata/TriggerType";

export default interface Trigger {

    type: TriggerType
    actionId: string
    condition: string
    eventName: string
    propertyName: string
    calledActionId: string
    timeoutMillis: number
    debounceMillis: number
    // For OnCustomEvent triggers: where to listen and, for COMPONENT scope, the source to match.
    source?: 'DOCUMENT' | 'COMPONENT' | 'SELF'
    from?: string
    // OnLoad/OnSuccess: run the triggered action as a silent background refresh (no loading veil).
    background?: boolean

}

File: backend/shared/uidl/src/main/java/io/mateu/uidl/annotations/Trigger.java
package io.mateu.uidl.annotations;

import java.lang.annotation.Repeatable;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

@Repeatable(Triggers.class)
@Retention(RetentionPolicy.RUNTIME)
public @interface Trigger {

  TriggerType type();

  String actionId();

  int timeoutMillis() default 0;

  int times() default 1;

  String condition() default "";

  String calledActionId() default "";

  String propertyName() default "";

  String eventName() default "";

  /**
   * Run the triggered action as a silent background refresh — no loading veil, no busy affordance.
   * For {@code OnLoad}/{@code OnSuccess} status polls that re-fetch a screen in place; leave false
   * (the default) for a first-load fetch that should show the veil.
   */
  boolean background() default false;
}
